src/librustc_trait_selection/traits/error_reporting/mod.rs

```diff
@@ -316,10 +316,11 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                         );
 
                         let should_convert_option_to_result =
-                            format!("{}", trait_ref.print_only_trait_path())
+                            format!("{}", trait_ref.print_only_trait_path_full())
                                 .starts_with("std::convert::From<std::option::NoneError");
-                        let should_convert_result_to_option = format!("{}", trait_ref)
-                            .starts_with("<std::option::NoneError as std::convert::From<");
+                        let should_convert_result_to_option =
+                            format!("{}", trait_ref.print_full_path())
+                                .starts_with("<std::option::NoneError as std::convert::From<");
```

Comment:
`NoneError` should be a diagnostic item, instead of string matching.

Reply:
Addressed in https://github.com/rust-lang/rust/pull/74344

Reply:
Thanks to #74344, changes to `src/librustc_trait_selection/traits/error_reporting/mod.rs` reverted in this PR.